File: results/RQ1/05_baseline_qualification/units/p07_calculator/java/accepted/repository/src/main/java/Calculator.java
/**
 * Canonical public API for the calculator.
 * Coordinates Tokenizer and Parser, wraps outcomes into CalculationResult,
 * manages CalculationHistory bounded storage, and exposes history() and clearHistory().
 */
public class Calculator {
    /** The maximum capacity of the calculation history */
    private final int historyCapacity;
    
    /** The history store for calculations */
    private final CalculationHistory history;

    /**
     * Constructs a new Calculator with the default history capacity of 10.
     */
    public Calculator() {
        this(10);
    }

    /**
     * Constructs a new Calculator with the specified history capacity.
     *
     * @param historyCapacity the maximum number of entries to store in history
     */
    public Calculator(int historyCapacity) {
        this.historyCapacity = historyCapacity;
        this.history = new CalculationHistory(historyCapacity);
    }

    /**
     * Evaluates the given arithmetic expression.
     * 
     * @param expression the arithmetic expression to evaluate
     * @return a CalculationResult containing either the result or an error message
     */
    public CalculationResult evaluate(String expression) {
        try {
            // Tokenize the expression
            Tokenizer tokenizer = new Tokenizer(expression);
            java.util.List<Token> tokens = tokenizer.tokenize();
            
            // Parse and evaluate
            Parser parser = new Parser(tokens);
            double result = parser.evaluate();
            
            // Store in history
            HistoryEntry entry = new HistoryEntry(expression, result);
            history.add(entry);
            
            return new CalculationResult(result);
        } catch (CalculatorException e) {
            // Store in history
            HistoryEntry entry = new HistoryEntry(expression, e.getMessage());
            history.add(entry);
            
            return new CalculationResult(e.getMessage());
        }
    }

    /**
     * Returns the calculation history.
     * 
     * @return a list of HistoryEntry objects in insertion order
     */
    public java.util.List<HistoryEntry> history() {
        return history.getAll();
    }

    /**
     * Clears all entries from the calculation history.
     */
    public void clearHistory() {
        history.clear();
    }

    /**
     * Returns the maximum capacity of the calculation history.
     * 
     * @return the history capacity
     */
    public int historyCapacity() {
        return historyCapacity;
    }
}

File: results/RQ1/05_baseline_qualification/units/p07_calculator/cpp/replays/0001_seed_3978176900/p07_calculator_cpp/p07_calculator_cpp/src/Calculator.cpp
#include "Calculator.h"
#include "Lexer.h"
#include "Parser.h"
#include "Evaluator.h"

namespace calc {

Calculator::Calculator() = default;

EvaluationOutcome Calculator::evaluate(const std::string& expression) {
    // Tokenize the input expression
    Lexer lexer;
    LexResult lexResult = lexer.tokenize(expression);
    
    if (!lexResult.success) {
        return {false, 0.0, lexResult.errorMessage, lexResult.errorPosition};
    }
    
    // Parse tokens into RPN
    Parser parser;
    ParseResult parseResult = parser.toRPN(lexResult.tokens);
    
    if (!parseResult.success) {
        return {false, 0.0, parseResult.errorMessage, parseResult.errorPosition};
    }
    
    // Evaluate the RPN expression
    Evaluator evaluator;
    EvalResult evalResult = evaluator.evaluateRPN(parseResult.rpn);
    
    if (!evalResult.success) {
        return {false, 0.0, evalResult.errorMessage, evalResult.errorPosition};
    }
    
    // Record successful evaluation in history
    history_.record(expression, evalResult.value);
    
    return {true, evalResult.value, "", 0};
}

const std::vector<HistoryEntry>& Calculator::history() const {
    return history_.entries();
}

void Calculator::clearHistory() {
    history_.clear();
}

} // namespace calc